test/e2e/upgrade_hypershift_operator_test.go

```diff
@@ -0,0 +1,226 @@
+			t.Log("Waiting for nodes to report ready")
+			_ = e2eutil.WaitForGuestClient(t, ctx, mc, hc)
+
+			t.Log("Nodes are ready")
+			mgmtClient = mc
+			hostedCluster = hc
+
+			g.Expect(t.Run("Calculate HyperShift Operator upgrade invariants", func(t *testing.T) {
+				t.Log("Calculating HyperShift Operator upgrade invariants")
+				g := gomega.NewWithT(t)
+
+				// Get the newly created default NodePool
+				nodepools := &hyperv1.NodePoolList{}
+				err = mgmtClient.List(ctx, nodepools, crclient.InNamespace(hostedCluster.Namespace))
+
+				g.Expect(err).ToNot(gomega.HaveOccurred(), "Listing nodepools in namespace %s shouldn't return errors",
+					hostedCluster.Namespace)
+				g.Expect(nodepools.Items).ToNot(gomega.BeEmpty(), "Should find NodePools in namespace %s",
+					hostedCluster.Namespace)
+
+				nodePoolsMap = make(map[string]*hyperv1.NodePool, len(nodepools.Items))
+				t.Logf("Found %d NodePools", len(nodepools.Items))
+				for i := range nodepools.Items {
+					t.Logf("Found NodePool %s", nodepools.Items[i].Name)
+					nodePoolsMap[nodepools.Items[i].Name] = &nodepools.Items[i]
+				}
+
+				// Get the MachineDeployments
+				machineDeployments := &v1beta1.MachineDeploymentList{}
+				hcpNameSpace = manifests.HostedControlPlaneNamespace(hostedCluster.Namespace, hostedCluster.Name)
+
+				err = mgmtClient.List(ctx, machineDeployments, crclient.InNamespace(hcpNameSpace))
+
+				g.Expect(err).ToNot(gomega.HaveOccurred(),
+					"Listing MachineDeployments in namespace %s shouldn't return errors", hcpNameSpace)
+				g.Expect(machineDeployments.Items).ToNot(gomega.BeEmpty(),
+					"Should find MachineDeployments in namespace %s", hcpNameSpace)
+				g.Expect(len(machineDeployments.Items)).To(gomega.BeEquivalentTo(len(nodepools.Items)),
+					"Number of MachineDeployments and NodePools should match")
+
+				machineDeploymentMap = make(map[string]*v1beta1.MachineDeployment)
+				t.Logf("Found %d MachineDeployments", len(machineDeployments.Items))
+				for i := range machineDeployments.Items {
+					t.Logf("Found MachineDeployment %s", machineDeployments.Items[i].Name)
+					machineDeploymentMap[machineDeployments.Items[i].Name] = &machineDeployments.Items[i]
+				}
+
+			})).To(gomega.BeTrue(), "Calculating HyperShift Operator upgrade invariants should succeed")
+
+			g.Expect(t.Run("Upgrade HyperShift Operator", func(t *testing.T) {
+				t.Logf("Upgrading HyperShift Operator to image %s", hyperShiftOperatorLatestImage)
+				globalOpts.HyperShiftOperatorLatestImage = hyperShiftOperatorLatestImage
+				// Note that we're replacing previous HO installation cleanup function with the new one
+				cleanup, err = e2eutil.InstallHyperShiftOperator(ctx, globalOpts)
+				if err != nil {
+					t.Fatal("Failed to upgrade HyperShift Operator")
+				}
+			})).To(gomega.BeTrue(), "Upgrade HyperShift Operator should succeed")
+
+			g.Expect(t.Run("Verify upgrade invariants", func(t *testing.T) {
```

Comment:
imo, it would be cleaner if the steps in the test `Calculate HyperShift Operator upgrade invariants` -> `Verify upgrade invariants` are not subtests, as you had to wrap them with `g.Expect()` which would get us double error if the subtest fail, and the fact that they share state like nodePoolMap and machineDeploymentMap.

no strong feelings though.

Reply:
I'm optimizing more for the test output readability, IMHO any test has three main steps, pre-condition, action, verification. 

So now the test output looks like this, which is IMHO is a cleaner and more readable output.

```
    --- PASS: TestUpgradeHyperShiftOperator/Main (16.71s)
        --- PASS: TestUpgradeHyperShiftOperator/Main/Calculate_HyperShift_Operator_upgrade_invariants (0.32s)
        --- PASS: TestUpgradeHyperShiftOperator/Main/Upgrade_HyperShift_Operator (15.77s)
        --- PASS: TestUpgradeHyperShiftOperator/Main/Verify_upgrade_invariants (0.22s)
```